### PR TITLE
The pull request for the part of gravity anomaly from SUSTech

### PR DESCRIPTION
**Pull request by wwzzyyzzrr**
This is Southern University of Science and Technology. We design and write a new part for this project about the gravity anomaly. We write the python file by the structure like the file in the project before, and the folder structure is also like the project structure.

### CODE REVIEW

**Comment by lheagy** [comment]
👋 Hi @thast would you be willing to review the notebook added by @wwzzyyzzrr?

**Comment by thast** [comment]
on my to-do list.

**Comment by lheagy** [comment]
many thanks @thast :) Please let me know if you need anything from me

**Comment by thast** [comment]
Hi @wwzzyyzzrr,

Thanks a lot for your first-time contribution to this repository, that's awesome! Gravity is a really key feature and many people will have use for it.

I want first to highlight a couple of points here before doing any in-depth review. This will help get things right from the start and the long-term use of the work you have developed:
- SimPEG.PF have been fixed. Your PR is failing on the style test (see the  [contributing page](https://github.com/geoscixyz/geosci-labs/blob/master/CONTRIBUTING.md) for more details). Having consistent formatting help the readability. I will start the review once Travis passed. If you experience difficulties with it, please feel free to contact us and we will help.
- I notice there is no comment at all in your code. Please add some documentation to your functions and code. This will help both the review and the maintenance, as this is likely to be used, improved and built upon in the future.
- Once the PR is merged, we also usually release PR notes, basically a description of what have been added, fixed or improved upon (like names of the new notebooks, what do they do and how etc.). If you could provide it, that will also help both the review and the users.

For all those points, please take a look at the [contributing page](https://github.com/geoscixyz/geosci-labs/blob/master/CONTRIBUTING.md) for more details (like on PEP8, Black, Flake8 for formatting, PR description etc.) 

Please feel free to contact us and ask any question you might have.

@lheagy : let us know if you have any additional comment or thought about it.

**Comment by wwzzyyzzrr** [comment]
Hi, @Thibaut Astic:
     Thankyou for your work. We will finsh it in time and pull request
again. And for the comments, we had found that the codes in the folder
geoscilabs have no comment, so we delete all of them, what should we do for
it next step? Add comments in the python file or write in document?


2019年2月11日 09:57，"Thibaut Astic" <notifications@github.com>写道：

Hi @wwzzyyzzrr <https://github.com/wwzzyyzzrr>,

Thanks a lot for your first-time contribution to this repository, that's
awesome! Gravity is a really key feature and many people will have use for
it.

I want first to highlight a couple of points here before doing any in-depth
review. This will help get things right from the start and the long-term
use of the work you have developed:

   - SimPEG.PF have been fixed. Your PR is failing on the style test (see
   the contributing page
   <https://github.com/geoscixyz/geosci-labs/blob/master/CONTRIBUTING.md>
   for more details). Having consistent formatting help the readability. I
   will start the review once Travis passed. If you experience difficulties
   with it, please feel free to contact us and we will help.
   - I notice there is no comment at all in your code. Please add some
   documentation to your functions and code. This will help both the review
   and the maintenance, as this is likely to be used, improved and built upon
   in the future.
   - Once the PR is merged, we also usually release PR notes, basically a
   description of what have been added, fixed or improved upon (like names of
   the new notebooks, what do they do and how etc.). If you could provide it,
   that will also help both the review and the users.

For all those points, please take a look at the contributing page
<https://github.com/geoscixyz/geosci-labs/blob/master/CONTRIBUTING.md> for
more details (like on PEP8, Black, Flake8 for formatting, PR description
etc.)

Please feel free to contact us and ask any question you might have.

@lheagy <https://github.com/lheagy> : let us know if you have any
additional comment or thought about it.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/geoscixyz/geosci-labs/pull/12#issuecomment-462203018>,
or mute the thread

**Comment by thast** [comment]
@wwzzyyzzrr : Ah I see! Sorry for the confusion. Documentation and comments in the python file is better in my opinion, makes it easier to understand, review, debug and improve. I do not know how easy it is for you to "put it back" but that would be really valuable. Also from a long-term perspective, we would likely build the API documentation for the code comments itself.

I think that the codes that are still in geosci-labs were  developed years ago, before it became a shared resources, thus a lack of time dedicated to document it. The most advanced programs got moved to SimPEG and documented, but the ones really more specific to the original teaching material stayed poorly documented.

**Comment by lheagy** [comment]
Many thanks @thast for leading the review! Just one minor thing I will jump in with, the test that is failing is the "style" test, meaning that the code is not formatted. From the main directory (geosci-labs) you can run:
```
make format
```
to format the code (or have a look at the Makefile and copy the command into your terminal). This runs [black](https://github.com/ambv/black) so that the code if formatted consistently. You can then commit and push any changes it suggests. 

After this, try running
```
make check
```
and this will check that the code has been formatted and will also check that the code style (loosley) follows the [pep8 guidelines](https://www.python.org/dev/peps/pep-0008/) for naming and a few other best-practices. If `make check` runs cleanly, then the test should pass on travis.

**Comment by wwzzyyzzrr** [comment]
I have modified the files that I think need to be changed. And the test of the flake8 is also pass, so could you review the commits of us again? Thank you.

**Comment by thast** [comment]
Good job @wwzzyyzzrr on the style repair :)
It seems the gravity notebooks are not tested in Travis (one travis test per module, and I do not see gravity in the list). Could you add the test? Let us know if you need help.
image

**Comment by wwzzyyzzrr** [comment]
Thanks, @thast :
  We have added the test file for the gravity and add it into the .travis.yml, all the tests have been passed.

File: environment.yml [hunk 1]
```diff
@@ -17,4 +17,4 @@ dependencies:
     - deepdish
     - SimPEG
     - empymod
-    - git+https://github.com/geoscixyz/geosci-labs
+    - git+https://github.com/wwzzyyzzrr/geosci-labs
```
Comment (lheagy) [inline environment.yml]: could you please revert this change? the environment.yml should point to the main repository.

File: geoscilabs/gravity/gravityDike.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    print("{0: ^10}{1: ^10}".format("X", "Δgz"))
+    for i in range(len(currentResult[0])):
+        print(
+            "{0: ^10}{1: ^10}".format(
+                "%.3f" % currentResult[0][i], "%.6f" % currentResult[1][i]
+            )
+        )
+
+
+if __name__ == "__main__":
```
Comment (lheagy) [inline geoscilabs/gravity/gravityDike.py]: this doesn't need to be here. It is really only necessary if you expect people to try and run this file as a python script, but users will be interacting with it by importing the package.

File: requirements.txt [hunk 3]
```diff
@@ -9,4 +9,4 @@ ipywidgets
 deepdish
 Pillow
 requests
-git+https://github.com/geoscixyz/geoscilabs
+git+https://github.com/wwzzyyzzrr/geosci-labs
```
Comment (lheagy) [inline requirements.txt]: same comment here, please use the geoscixyz version, not the fork

File: setup.py [hunk 4]
```diff
@@ -51,7 +51,7 @@
     long_description=LONG_DESCRIPTION,
     keywords="geophysics, electromagnetics",
     url="https://geosci.xyz",
-    download_url="https://github.com/geoscixyz/geoscilabs",
+    download_url="https://github.com/wwzzyyzzrr/geosci-labs",
```
Comment (lheagy) [inline setup.py]: and again here. Please stick with the geoscixyz version

File: environment.yml (hunk 1 above)
Comment (lheagy) [reply environment.yml]: In the future, you can just create a branch on this repo rather than using a fork and that should simplify things a bit. You should now have the ability to push to this repo since you are a geoscixyz team member 🎉

**Comment by wwzzyyzzrr** [comment]
Hi, @lheagy :
     Thanks for your suggestions, I have revised the problems mentioned above.

File: geoscilabs/gravity/gravityDike.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    y = x.copy()
+    X, Y = np.meshgrid(x, y)
+    for each in range(npts):
+        gravity_change.append(calculategravity(delta_rho, z1, z2, b, beta, x[each]))
+    return respEW, gravity_change, X, Y
+
+
+# get the value of the delta gravity
+def calculategravity(delta_rho, z1, z2, b, beta, x) -> float:
+    G = 6.67259e-8
```
Comment (thast) [inline geoscilabs/gravity/gravityDike.py]: I think it would be better here to add in the import:
`from scipy.constants import G` (be careful with the units)

File: setup.py [hunk 6]
```diff
@@ -51,7 +51,7 @@
     long_description=LONG_DESCRIPTION,
     keywords="geophysics, electromagnetics",
     url="https://geosci.xyz",
-    download_url="https://github.com/geoscixyz/geoscilabs",
+    download_url="https://github.com/geoscixyz/geosci-labs",
```
Comment (thast) [inline setup.py]: good catch!

File: geoscilabs/gravity/gravityDike.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    "midnightblue",
+    "yellow",
+    "gold",
+    "lime",
+]  # The Colors
+index = 0
+currentResult = []
+
+# The main function and draw the first table
+def drawfunction(delta_rho, z1, z2, b, beta, stationSpacing, B):
```
Comment (thast) [inline geoscilabs/gravity/gravityDike.py]: It would be good to have documentation such as:
`def forward(nD,delta_rho):
"""
   Parameters
        ----------
        nD: int
            number of data
        delta_rho : numpy.array
            model of density contrasts

        Returns
        -------
        np.array
            data, the forward data
"""

But maybe not required for this PR. What's your opinion on it @lheagy ?

**Comment by thast** [comment]
Hi @wwzzyyzzrr , thanks for your work. I have left  a few comments above for you to review.  If @lheagy also agree to let the documentation of the functions as it is, then switching the hard coded value of G for the ` import scipy.constants import G` (with proper unit conversion) will do for the current PR. Please also provide releases notes: basically a description of what have been added, fixed or improved upon (like names of the new notebooks, what do they do and how etc.), to accompany the official release. It does not need to be long, just a few lines will do.
Good job!

File: geoscilabs/gravity/gravityDike.py (hunk 7 above)
Comment (lheagy) [reply geoscilabs/gravity/gravityDike.py]: It would be nice and very much appreciated, but I don't think we can ask that it be required for this PR. We haven't set that precedent within the codebase yet.

Comment (thast) [reply geoscilabs/gravity/gravityDike.py]: Sounds good!
